pyannote/audio/utils/torchmetrics.py

```diff
@@ -0,0 +1,240 @@
+    preds, target = torch.transpose(preds, 1, 2), torch.transpose(target, 1, 2)
+
+    batch_size, num_samples, num_classes_1 = target.shape
+    batch_size_, num_samples_, num_classes_2 = preds.shape
+    if (
+        batch_size != batch_size_
+        or num_samples != num_samples_
+        or num_classes_1 != num_classes_2
+    ):
+        msg = f"Shape mismatch: {tuple(target.shape)} vs. {tuple(preds.shape)}."
+        raise ValueError(msg)
+    return preds, target
+
+
+def compute_der_values(
+    preds: torch.Tensor, target: torch.Tensor, threshold: float
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """Compute the false alarm, missed detection, confusion and total values.
+
+    Parameters
+    ----------
+    preds : torch.Tensor
+        preds tensor of shape (B,F,C)
+    target : torch.Tensor
+        preds tensor of shape (B,F,C) (must only contain 0s and 1s)
+    threshold : float
+        threshold to discretize preds
+
+    Returns
+    -------
+    _type_
+        _description_
+    """
+
+    preds_bin = (preds > threshold).float()
+
+    hypothesis, _ = permutate(target, preds_bin)
+
+    detection_error = torch.sum(hypothesis, 2) - torch.sum(target, 2)
+    false_alarm = torch.maximum(detection_error, torch.zeros_like(detection_error))
+    missed_detection = torch.maximum(
+        -detection_error, torch.zeros_like(detection_error)
+    )
+
+    confusion = torch.sum((hypothesis != target) * hypothesis, 2) - false_alarm
+
+    false_alarm = torch.sum(false_alarm)
+    missed_detection = torch.sum(missed_detection)
+    confusion = torch.sum(confusion)
+    total = 1.0 * torch.sum(target)
+
+    return false_alarm, missed_detection, confusion, total
+
+
+class DER(Metric):
+    """
+    Compute Diarization Error Rate on discretized annotations.
+
+    Expects preds and target tensors of the shape (NUM_BATCH, NUM_CLASSES, NUM_FRAMES) in its update.
+
+    Note that this is only a reliable metric if num_frames == the total number of frames of the diarized audio.
+    """
+
+    def __init__(self, threshold: float = 0.5):
+        super().__init__()
+
+        self.threshold = threshold
+
+        self.add_state("false_alarm", default=torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state(
+            "missed_detection", default=torch.tensor(0.0), dist_reduce_fx="sum"
+        )
+        self.add_state("confusion", default=torch.tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("total", default=torch.tensor(0.0), dist_reduce_fx="sum")
+
+    def update(
+        self,
+        preds: torch.Tensor,
+        target: torch.Tensor,
+    ):
+        # switch back to pyannote's tensor shape : from (batch,class,...) to (batch,frames,class)
+        preds, target = der_tensors_setup(preds, target)
+
+        false_alarm, missed_detection, confusion, total = compute_der_values(
+            preds, target, self.threshold
+        )
```

Comment:
I'd rather do everything with torchmetrics expected shapes:
- [ ] remove `der_tensors_setup` reshaping
- [ ] update `compute_der_values` to expect torchmetrics shapes

Reply:
I kept pyannote shapes for `compute_der_values` because that's what `permutate` expects/outputs.
Should I directly do 
```python
hypothesis, _ = permutate(torch.transpose(target, 1, 2), torch.transpose(preds_bin, 1, 2))
hypothesis = torch.transpose(hypothesis, 1, 2)
```
instead ?

Reply:
Yes, something like that! 

The (long term) idea would be to make a PR to `torchmetrics` adding these metrics.
So it is better to stick as much as possible to `torchmetrics` shapes and conventions.